### PR TITLE
Parse the binary expression and eval statement

### PR DESCRIPTION
**Pull request by codingpotato**
- Parse number and binary expression, eval statement into AST tree.
- Suppress CS3021 warnings of ANTLR generated code.

### CODE REVIEW

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+namespace SeedLang.X {
+  // The visitor class to visit each SeedPython parse tree and generate the corresponding AST tree.
+  //
+  // The default implement of SeedPythonBaseVisitor is to visit all the children and return the
+  // result of the last one. PythonVisitor overrides the method if the default implement is not
+  // correct.
+  internal class PythonVisitor : SeedPythonBaseVisitor<AstNode> {
+    // Visits add and subtract binary expressions.
+    //
+    // The expr() method of the Add_subContext return a ExprContext array which contains exact 2
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:30]: return -> returns

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    // items: the left and right ExprContexts.
+    public override AstNode VisitAdd_sub([NotNull] SeedPythonParser.Add_subContext context) {
+      return BuildBinary(context.op, context.expr());
+    }
+
+    // Visits multiply and divide binary expressions.
+    //
+    // The expr() method of the Add_subContext return a ExprContext array which contains exact 2
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:38]: return -> returns

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Visits the eval statement.
+    public override AstNode VisitEval_stmt([NotNull] SeedPythonParser.Eval_stmtContext context) {
+      var expr = Visit(context.expr()) as Expression;
+      return Statement.Eval(expr);
+    }
+
+    // Builds a binary expression node from the opToken and exprs.
+    //
+    // The exprContexts parameter must contains exact 2 items: the left and right ExprContext.
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:78]: contains -> contain

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    // Visits number expression.
+    public override AstNode VisitNumber([NotNull] SeedPythonParser.NumberContext context) {
+      return Expression.Number(double.Parse(context.GetText()));
+    }
+
+    // Visits grouping expression.
+    //
+    // There isn't corresponding grouping AST node. The order of the expression node in the AST tree
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:51]: Either "There isn't any corresponding grouping AST node" or "There is no corresponding grouping AST node".

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    // represent the grouping structure.
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:52]: represent -> represents

File: grammars/SeedPython.g4 [hunk 6]
```diff
@@ -48,6 +48,10 @@
```
Comment (wixette) [inline grammars/SeedPython.g4:48]: Not related to this PR: it seems there is a trailing space at line #15. Please clear it. You may want to install a VS Code extension called "trailing Spaces" to check them.

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    public override AstNode VisitGrouping([NotNull] SeedPythonParser.GroupingContext context) {
+      return Visit(context.expr());
+    }
+
+    // Visits the simple statement.
+    //
+    // The small_stmt() method of the Simple_stmtContext returns a array which contains all the
+    // small statements. There is at least one small statement in it.
+    //
+    // TODO: parse all the small statements in it, only parse the first one now.
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:62]: It'd be better to move the TODO comments into the method body.

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Visits multiply and divide binary expressions.
+    //
+    // The expr() method of the Add_subContext return a ExprContext array which contains exact 2
+    // items: the left and right ExprContexts.
+    public override AstNode VisitMul_div([NotNull] SeedPythonParser.Mul_divContext context) {
+      return BuildBinary(context.op, context.expr());
+    }
+
+    // Visits number expression.
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:44]: expression -> expressions

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+    // Visits grouping expression.
+    //
+    // There isn't corresponding grouping AST node. The order of the expression node in the AST tree
+    // represent the grouping structure.
+    public override AstNode VisitGrouping([NotNull] SeedPythonParser.GroupingContext context) {
+      return Visit(context.expr());
+    }
+
+    // Visits the simple statement.
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:57]: the simple statement -> simple statements

File: csharp/src/SeedLang.X/PythonVisitor.cs [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    //
+    // TODO: parse all the small statements in it, only parse the first one now.
+    public override AstNode VisitSimple_stmt(
+      [NotNull] SeedPythonParser.Simple_stmtContext context) {
+      SeedPythonParser.Small_stmtContext[] smallStatements = context.small_stmt();
+      Debug.Assert(smallStatements.Length > 0);
+      return Visit(smallStatements[0]);
+    }
+
+    // Visits the eval statement.
```
Comment (wixette) [inline csharp/src/SeedLang.X/PythonVisitor.cs:70]: the eval statement. -> eval statements.

Please make the article usage consistent across all the method comments in this class. I think either "Visits a ... expression" or "Visits ... expressions" is okay here.

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 1 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:30]: Done

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 2 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:38]: Done

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 8 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:44]: Done

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 4 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:51]: Done

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 5 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:52]: Done

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 9 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:57]: Done

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 7 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:62]: Done

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 10 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:70]: Done

File: csharp/src/SeedLang.X/PythonVisitor.cs (hunk 3 above)
Comment (codingpotato) [reply csharp/src/SeedLang.X/PythonVisitor.cs:78]: Done

File: grammars/SeedPython.g4 (hunk 6 above)
Comment (codingpotato) [reply grammars/SeedPython.g4:48]: Done

Comment (codingpotato) [reply grammars/SeedPython.g4:48]: Done

**Review by wixette — APPROVED** [review]

**Pull request merged**